enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,169 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of OpenShift 4 is to make upgrades seamless and risk free.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give the operators ability to run some checks before the upgrade to find out if the new incoming version is compatible with the existing version.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before going through the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered check if upgrade to the target version is going to be successfully without doing an upgrade.
+Use preflight checks to improve the success rate of upgrades.
+
+
+### Goals
+
+* This enhancement proposes a framework for preflight checks and discusses the CVO side implementation of that framework.
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
+* Downloading and verification of images associated with release payload should be part of preflight check.
+
+### Non-Goals
+
+* This enhancement does not commit particular operators for providing preflight implementations or discuss the operator-specific logic that should be executed by preflight checks.
+* The preflight checks should not be used as a migration mechanism.
+* It is not the responsibility of preflight checks to protect against upgrades involving more than one minor version i.e. 4.y.z to 4.(y+2).z.
+
+## Proposal
+
+Operators need to provide a different bin path for the preflight checks i.e. /bin/preflight.
+The bin path need to exist inside of the operator image.
+Cluster version operator (CVO) will be running a single job per operator for the preflight checks.
+/bin/preflight must be executable. It can be symbolic link to the operator binary or a script or It can be non existent.
+Users can skip or override the preflight checks with `--force`.
+
+### User Stories
+
+#### Story 1
+
+Some customers would prefer to disable automatic credential provisioning by Cloud credential operator (CCO) and instead manage it themselves.
+However if new permissions are required by the target version, the upgrade could block midway even if the administrator provisioned more powerful credentials.
+With the help of preflight checks CCO can check if they can safely upgrade to a given release before doing the actual upgrade.
+
+### Implementation Details
+
+CVO will run the preflight check from the new operator image to which the current operator image is trying to upgrade.
```

Comment:
It would be useful if the preflight check could make use of the new _operand_ image.  We have a specific use case with cluster-ingress-operator where the operand is an ingress controller, which the operator configures using an administrator-provided default serving certificate.  We want to write a preflight check for cluster-ingress-operator that would use the new operand image to verify that it did not reject the current certificate in use.  

This use-case is inspired by the recent changeover to a RHEL8 base image, which brought a new OpenSSL version that tightened restrictions on TLS keys, which caused the new operand image to reject keys that the old operand image accepted ([BZ#1874278](https://bugzilla.redhat.com/show_bug.cgi?id=1874278)).

Absent the option to use the operand image, alternative approaches to writing such a pre-flight check would be to import the operand image (or the executable and libraries from it) when building the operator image, or writing a small application to approximate the operand's behavior as closely as possible, which would be more effort and be more error-prone than using the operand image directly.

Reply:
I don't think the CVO cares about this distinction.  You should be able to just reference whatever you need in the `image` property of your recommending Job.  We should probably pivot to more generic language to make that more clear.  In this case "new image" is probably sufficient.

Reply:
Right, It should depend on the job manifest you define as part of the  preflight check.

Reply:
Added this as an usecase in https://github.com/openshift/enhancements/pull/363/commits/4ca3ea806ece54f59b3760d00308992f9df512e3